# The story so far

You are here because you have a grasp of the R skills required to complete R Programming 1. See [here](https://sccbioinformatics.github.io/R_programming_1/) for the course material. In summary, that course covers:

* RStudio
  + Navigating the panes
  + Writing scripts
* Vectors
* Matrices
  + Subsetting
  + Manipulating
* Lists
  * Creating
  * Accessing
* Data Frames
  + Making ggplots from
* Reading and writing files
  + Reading in expression data
  + Writing out tables
* Flow control basics
  + `for`,`if`, `else` statements
* Functions
  + Creating
  + Using them
* `apply`
  + Using built-in functions like `sum` and `var`
  + Using `apply` with your own functions
* Standardising data
  + z-scoring data for clustering
* Clustering
  + Hierarchical clustering rows and columns
  + `pheatmap`
* Linear models
  + Writing the models
  + Extracting differentially expressed genes
* Kmeans
  + Programming from scratch
  + Starting to write algorithms
* The Final Project
  + Clustering using simulated annealing

If you aren't comfortable with most of this list, then this course might not be for you.

## What will will cover

This course will take it up a notch, and we will demonstrate all the concepts mentioned earlier using single-cell data as our primary working material. Using this, we will learn how data is organised in R and how functions are written to analyse this data by using Seurat as an example. We will also learn how to make a small R package.

Importantly, we will learn how to use version control and Github to make your code available online to anyone. Your package will installable from anywhere with an internet connection. Because version control is so important, we will start with this.


